[SPARK-21079] [SQL] Calculate total size of a partition table as a sum of individual partitions

mbasmanova opened the pull request:
## What changes were proposed in this pull request?

Storage URI of a partitioned table may or may not point to a directory under which individual partitions are stored. In fact, individual partitions may be located in totally unrelated directories. Before this change, ANALYZE TABLE table COMPUTE STATISTICS command calculated total size of a table by adding up sizes of files found under table's storage URI. This calculation could produce 0 if partitions are stored elsewhere.

This change uses storage URIs of individual partitions to calculate the sizes of all partitions of a table and adds these up to produce the total size of a table.

CC: @wzhfy

## How was this patch tested?

Added unit test.

Ran ANALYZE TABLE xxx COMPUTE STATISTICS on a partitioned Hive table and verified that sizeInBytes is calculated correctly. Before this change, the size would be zero.

mbasmanova committed:
[SPARK-21079] [SQL] Calculate total size of a partition table as a sum of individual partitions

AmplabJenkins commented:
Can one of the admins verify this patch?

tejasapatil commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -109,16 +124,16 @@ object AnalyzeTableCommand extends Logging {
       val path = new Path(p)
       try {
         val fs = path.getFileSystem(sessionState.newHadoopConf())
         calculateTableSize(fs, path)
```
given that after this change this would be either a table / partition, we should rename the method

tejasapatil replied:
looks like this comment got missed

tejasapatil commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -81,6 +83,19 @@ case class AnalyzeTableCommand(
+    if (catalogTable.partitionColumnNames.isEmpty) {
+      calculateTotalSize(sessionState, catalogTable.identifier, catalogTable.storage.locationUri)
+    } else {
+      // Table = Sum(partitions)
```
nit : you could explain it instead of putting out the formula

tejasapatil commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -81,6 +83,19 @@ case class AnalyzeTableCommand(
+    }
+  }
+
+  private def calculateTotalSize(sessionState: SessionState, tableId: TableIdentifier,
```
nit: if the number of args exceeds line length, the code style is to have individual args on separate line with 4 spaces.

```
  private def calculateTotalSize(
      sessionState: SessionState,
      tableId: TableIdentifier,
      locationUri: Option[URI]): Long = {
    // This method is mainly based on
    ....
```

wzhfy commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -81,6 +83,19 @@ case class AnalyzeTableCommand(
 
   def calculateTotalSize(sessionState: SessionState, catalogTable: CatalogTable): Long = {
+    if (catalogTable.partitionColumnNames.isEmpty) {
+      calculateTotalSize(sessionState, catalogTable.identifier, catalogTable.storage.locationUri)
```
rename `calculateLocationSize`?

wzhfy commented:
Please explain the failed scenario in PR description. That's why our test cases cannot catch it and we need to test it manually.

wzhfy commented:
Can you add a test case? In the test, we can add partitions with different paths by ALTER TABLE SET LOCATION command. I think that can reproduce your scenario, right?

mbasmanova committed:
[SPARK-21079][SQL] review comments + test

mbasmanova commented:
@wzhfy, @tejasapatil, thanks for your feedback. I think I addressed all of your comments and added a test. Please, take another look.

tejasapatil commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +128,40 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+        |INSERT INTO TABLE analyzeTable_part PARTITION (ds='2010-01-02')
+        |SELECT * FROM src
+      """.stripMargin).collect()
+    sql(
```
All the three partitions point reside under the same base dir. In reality it might be possible that all three are pointing to different locations. You could simulate that by creating empty partition (`ALTER TABLE ... ADD PARTITION`), then doing a `ALTER TABLE table_name [PARTITION partition_spec] SET LOCATION "new location"` and finally inserting data to the partition.

mbasmanova replied:
This is a fine suggestion, but I feel it falls outside of the scope of this change. This particular fix concerns only the case when table-level location is not a parent directory of a partition-level location.

wzhfy replied:
I agree with @tejasapatil . The test case would be more robust in that way.

mbasmanova commented:
@rxin, could you help unblock Jenkins for this PR?

wzhfy commented:
ok to test

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +128,40 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+      """.stripMargin).collect()
+
+    // Modify table location to not match location of individual partitions
+    sql("ALTER TABLE analyzeTable_part SET LOCATION 'file:/do/not/use'").collect()
```
please use `withTempDirs` to get several tmp dirs (deletion is guaranteed in `withTempDirs`)

wzhfy replied:
BTW, we need to move `withTempDirs` from `FileStreamSourceSuite` to `SQLTestUtils`.

wzhfy commented:
test this please

wzhfy commented:
Please also update description of PR, the key point of the problem is URI for table is different from that of its partitions, otherwise, it works fine for partitioned tables.

wzhfy commented:
Jenkins test this please

wzhfy commented:
@mbasmanova Let's first fix the above comments and ping admins to unblock this pr :)

mbasmanova committed:
[SPARK-21079][SQL] Rename a function + enhanced test

mbasmanova commented:
@wzhfy , @tejasapatil , I updated description of this PR, enhanced the test and renamed a function I missed before in Tejas's comment. How does it look now?

wzhfy commented on sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala:
```diff
@@ -126,6 +126,35 @@ private[sql] trait SQLTestUtils
+    } finally {
+      paths.foreach(path => Utils.deleteRecursively(path))
+    }
+  }
```
I personally think It's strange to use a recursive function, does the following modified function work for your test?
```

  protected def withTempDirs(num: Int)(body: Seq[File] => Unit) {
    val files = mutable.Buffer[File]()
    for (i <- 0 until num) {
      files += Utils.createTempDir().getCanonicalFile
    }
    try {
      body(files)
    } finally {
      files.foreach(Utils.deleteRecursively)
    }
  }
```

wzhfy commented on sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala:
```diff
@@ -126,6 +126,35 @@ private[sql] trait SQLTestUtils
   /**
+   * Generates the requested number of temporary paths without creating the
+   * actual files/directories and passes these to the provided function 'f'.
+   * Deletes any files/directories left at these paths after 'f' returns.
```
Creates the requested number of temporary directories, which are then passed to `f` and will be deleted after `f` returns.

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,48 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+      spark.table(tableName).queryExecution.analyzed.stats(conf).sizeInBytes
+
+    withTempPaths(4) { (paths) => paths match {
+      case tablePath :: partitionPaths =>
```
I think `withTempPaths(4) { paths =>` is ok, the pattern match seems unnecessary?

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,48 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+          """.stripMargin).collect()
+
+        val partitionDates = List("2010-01-01", "2010-01-02", "2010-01-03")
+        partitionDates.zip(partitionPaths).foreach(p => {
```
partitionDates.zip(partitionPaths).foreach { case (ds, path) =>

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,48 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+    def queryTotalSize(tableName: String): BigInt =
+      spark.table(tableName).queryExecution.analyzed.stats(conf).sizeInBytes
+
+    withTempPaths(4) { (paths) => paths match {
```
4 -> 3?

mbasmanova replied:
I need a path for the table and one more path for each of the three partitions. 1 + 3 = 4.

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,48 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+
+        assert(queryTotalSize("analyzeTable_part") === BigInt(17436))
+
+        sql("DROP TABLE analyzeTable_part").collect()
```
use `withTable` to include the test, then we don't need to drop table explicitly.

mbasmanova committed:
[SPARK-21079][SQL] incorporated comments about the test

mbasmanova commented:
@wzhfy, thank you careful review and helpful suggestions. I think I incorporated all of your comments. Would you take another look?

wzhfy commented on sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala:
```diff
@@ -126,6 +127,27 @@ private[sql] trait SQLTestUtils
   }
 
   /**
+   * Creates the requested number of temporary path (without creating the actual file/directory),
```
> without creating the actual file/directory

what do you mean by "actual file"?

mbasmanova replied:
I wanted to create a multi-path version of an existing withTempPath function. That function returns a *valid*, but non-existent path. The comments and path.delete() part of the implementation came from there.

/**
   * Generates a temporary path without creating the actual file/directory, then pass it to `f`. If
   * a file/directory is created there by `f`, it will be delete after `f` returns.
   *
   * @todo Probably this method should be moved to a more general place
   */
  protected def withTempPath(f: File => Unit): Unit = {
    val path = Utils.createTempDir()
    path.delete()
    try f(path) finally Utils.deleteRecursively(path)
  }

wzhfy commented on sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala:
```diff
@@ -126,6 +127,27 @@ private[sql] trait SQLTestUtils
+    val paths = mutable.ListBuffer[File]()
+    for (i <- 0 until num) {
+      val path = Utils.createTempDir().getCanonicalFile
+      path.delete()
```
why do we need to delete here?

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,45 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+
+    val tableName = "analyzeTable_part"
+    withTable(tableName) {
+      withTempPaths(4) {
```
add a line of comment to explain the number 4

gatorsmile replied:
`numPaths` -> `numPaths  = 4`

wzhfy commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,45 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+                   |ALTER TABLE ${tableName} ADD PARTITION (ds='${ds}')
+                   |LOCATION '${path.toString}'
+                """.
+                  stripMargin).collect()
```
nit: """.stripMargin)

wzhfy commented:
LGTM except some minor comments.
ping @cloud-fan @gatorsmile Could you start Jenkins for this pr?

gatorsmile commented:
ok to test

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala:
```diff
@@ -81,6 +83,21 @@ case class AnalyzeTableCommand(
+    if (catalogTable.partitionColumnNames.isEmpty) {
+      calculateLocationSize(sessionState, catalogTable.identifier, catalogTable.storage.locationUri)
+    } else {
+      // Calculate table size as a sum of its partitions. See SPARK-21079
```
`its partitions` -> `the visible partitions`

SparkQA commented:
**[Test build #78231 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78231/testReport)** for PR 18309 at commit [`b290bf9`](https://github.com/apache/spark/commit/b290bf9cd5f31eefa516f8185d01acba57ba4ce8).

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/test/SQLTestUtils.scala:
```diff
@@ -126,6 +127,27 @@ private[sql] trait SQLTestUtils
+   * Creates the requested number of temporary path (without creating the actual file/directory),
+   * which are then passed to f and will be deleted after f returns.
+   *
+   * @param num Number of directories to create
```
`num` -> `numPaths`

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,45 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+
+          sql(s"ANALYZE TABLE ${tableName} COMPUTE STATISTICS noscan")
+
+          assert(queryTotalSize(tableName) === BigInt(17436))
```
It can be simplified to 
```Scala
    val tableName = "analyzeTable_part"
    withTable(tableName) {
      withTempPath { path =>
        sql(
          s"""
             |CREATE TABLE $tableName (key STRING, value STRING) PARTITIONED BY (ds STRING)
             |LOCATION '$path'
           """.
            stripMargin)

        val partitionDates = List("2010-01-01", "2010-01-02", "2010-01-03")
        partitionDates.foreach { ds =>
          sql(
            s"""
               |INSERT INTO TABLE $tableName PARTITION (ds='$ds')
               |SELECT * FROM src
             """.stripMargin)
        }

        sql(s"ANALYZE TABLE $tableName COMPUTE STATISTICS noscan")

        assert(
          spark.table(tableName).queryExecution.analyzed.stats(conf).sizeInBytes === BigInt(17436))
      }
    }
```

gatorsmile replied:
After the test case simplification, the new utility function `withTempPaths` is not needed

mbasmanova replied:
This is how I wrote the test initially. In this case all partitions are located under the same top-level directory, but table-level location is somewhere else. I modified the test to use different paths for each partition as well as the table to address some of the earlier comments.

gatorsmile replied:
We do not need such complex test cases for verifying the fix by this PR. In the future, if needed, we can add this

mbasmanova replied:
Ok. I'll simplify the test and remove withTempPaths function. Since I need to make sure that partition locations are not under table location, I'll include ALTER TABLE $tableName SET LOCATION '$path' after creating table with partitions in the default location.

gatorsmile commented:
Could you add another test case? First, creates an external partitioned data source table (with an external location). Second, add only a subset of partitions. Third, analyze table but the statistics only include the visible partitions.

SparkQA commented:
**[Test build #78231 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78231/testReport)** for PR 18309 at commit [`b290bf9`](https://github.com/apache/spark/commit/b290bf9cd5f31eefa516f8185d01acba57ba4ce8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78231/
Test PASSed.

mbasmanova committed:
[SPARK-21079][SQL] addressed further comments

SparkQA commented:
**[Test build #78250 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78250/testReport)** for PR 18309 at commit [`09c4900`](https://github.com/apache/spark/commit/09c4900f2a4502b8e3d577798245b5cc3e29bb49).

mbasmanova commented:
@gatorsmile, re: additional test, how can I create a table with a mix of visible and invisible partitions?

SparkQA commented:
**[Test build #78250 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78250/testReport)** for PR 18309 at commit [`09c4900`](https://github.com/apache/spark/commit/09c4900f2a4502b8e3d577798245b5cc3e29bb49).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78250/
Test PASSed.

gatorsmile commented:
@mbasmanova 
```
create table ${fullTableName} using parquet options (path "${fullPath}")
```

If the path points to the directory of a partitioned table, the partitions are not visible unless you insert/load the data to that partition or you explicitly run `ALTER TABLE ADD PARTITION`.

mbasmanova committed:
[SPARK-21079][SQL] added test for a table with only a subset of visible partitions

mbasmanova commented:
@gatorsmile, thanks for clarifying. I added a test to analyze a table with only some of partitions being visible. Would you take another look?

SparkQA commented:
**[Test build #78292 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78292/testReport)** for PR 18309 at commit [`70f569b`](https://github.com/apache/spark/commit/70f569b9349115adc4e3a33c9fb100c9ce93b505).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78292/
Test FAILed.

shaneknapp commented:
i'll re-trigger this test once jenkins is back up.

mbasmanova commented:
Thank you, @shaneknapp .

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #78298 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78298/testReport)** for PR 18309 at commit [`70f569b`](https://github.com/apache/spark/commit/70f569b9349115adc4e3a33c9fb100c9ce93b505).

SparkQA commented:
**[Test build #78298 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78298/testReport)** for PR 18309 at commit [`70f569b`](https://github.com/apache/spark/commit/70f569b9349115adc4e3a33c9fb100c9ce93b505).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78298/
Test FAILed.

mbasmanova commented:
@shaneknapp, I can't think of why PySpark unit tests failed other than flakiness of the tests or infrastructure. Any chance you could trigger a re-run?

shaneknapp commented:
test this please

shaneknapp commented:
crap, this clobbered the pypy install.  fixing this now.

SparkQA commented:
**[Test build #78328 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78328/testReport)** for PR 18309 at commit [`70f569b`](https://github.com/apache/spark/commit/70f569b9349115adc4e3a33c9fb100c9ce93b505).

shaneknapp commented:
done

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78328/
Test FAILed.

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #78343 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78343/testReport)** for PR 18309 at commit [`70f569b`](https://github.com/apache/spark/commit/70f569b9349115adc4e3a33c9fb100c9ce93b505).

SparkQA commented:
**[Test build #78343 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78343/testReport)** for PR 18309 at commit [`70f569b`](https://github.com/apache/spark/commit/70f569b9349115adc4e3a33c9fb100c9ce93b505).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78343/
Test FAILed.

shaneknapp commented:
test this please

SparkQA commented:
**[Test build #78359 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78359/testReport)** for PR 18309 at commit [`70f569b`](https://github.com/apache/spark/commit/70f569b9349115adc4e3a33c9fb100c9ce93b505).

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,93 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+    val tableName = "analyzeTable_part_vis"
+    withTable(sourceTableName, tableName) {
+      withTempPath {
+        path =>
```
Nit: combine line 176 and 177

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,93 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+
+          val partitionDates = List("2010-01-01", "2010-01-02", "2010-01-03")
+          partitionDates.foreach {
+            ds =>
```
combine line 187 and 188

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,93 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+            ds =>
+              sql(
+                s"""
+                   |INSERT INTO TABLE ${sourceTableName} PARTITION (ds='${ds}')
```
`${sourceTableName}` -> `$sourceTableName`
`${ds}` -> `$ds`

Please fix the other similar issues

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,93 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+                s"""
+                   |INSERT INTO TABLE ${sourceTableName} PARTITION (ds='${ds}')
+                   |SELECT * FROM src
+                 """.stripMargin).collect()
```
No need to call `collect()`

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,93 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+               |CREATE TABLE ${tableName} (key STRING, value STRING)
+               |PARTITIONED BY (ds STRING)
+               |LOCATION '${path}'
+             """.stripMargin).collect()
```
The same here

gatorsmile commented:
Except the comments about test cases, the other parts look good to me.

SparkQA commented:
**[Test build #78359 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78359/testReport)** for PR 18309 at commit [`70f569b`](https://github.com/apache/spark/commit/70f569b9349115adc4e3a33c9fb100c9ce93b505).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78359/
Test PASSed.

mbasmanova committed:
[SPARK-21079][SQL] simplified first test; addressed comments for the second test

SparkQA commented:
**[Test build #78386 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78386/testReport)** for PR 18309 at commit [`c17917a`](https://github.com/apache/spark/commit/c17917a85ea91dfb29c842615735156a47eb008c).

mbasmanova committed:
[SPARK-21079][SQL] removed unnecessary import from SQLTestUtils.scala

mbasmanova commented:
@gatorsmile, thank you for review. All the comments should be addressed now. Please, check it out.

SparkQA commented:
**[Test build #78387 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78387/testReport)** for PR 18309 at commit [`05dacbc`](https://github.com/apache/spark/commit/05dacbc243cd34e3547572fbbd40ca31e33db30e).

SparkQA commented:
**[Test build #78386 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78386/testReport)** for PR 18309 at commit [`c17917a`](https://github.com/apache/spark/commit/c17917a85ea91dfb29c842615735156a47eb008c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78386/
Test PASSed.

SparkQA commented:
**[Test build #78387 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78387/testReport)** for PR 18309 at commit [`05dacbc`](https://github.com/apache/spark/commit/05dacbc243cd34e3547572fbbd40ca31e33db30e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/78387/
Test PASSed.

gatorsmile commented:
LGTM

gatorsmile commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala:
```diff
@@ -128,6 +129,77 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
+
+          // Analyze partial-copy table - expect only 1 partition
+          sql(s"ANALYZE TABLE $tableName COMPUTE STATISTICS noscan")
+          assert(queryTotalSize(tableName) === BigInt(5812))
```
I am afraid this hard-coded values might not succeed in some other platforms. We might need to adjust the way. However, let me first merge this. If needed, we might need to submit a follow-up PR. cc @cloud-fan @wzhfy

gatorsmile commented:
Thanks! Merging to master/2.2

gatorsmile commented:
Could you submit a backport PR to 2.1? Thanks!

maropu commented:
@gatorsmile This failure is related to this change? https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/78575/console

wzhfy commented:
@maropu yes, I've submitted a fix [here](https://github.com/apache/spark/pull/18415).

maropu commented:
oh, Thanks!

mbasmanova commented:
Folks, thanks for merging this PR. 

@gatorsmile, re: a request to backport to 2.1. Is this something that happens automatically? If not, could you point me to some instructions on how to do that?
